### DISCUSSION TITLE
String support

### DISCUSSION THREAD

**Discussion by franceccon** [OpenZeppelin/cairo-contracts#25] [Design]
When discussing #9 we sort of ignored that implementing `tokenUri` requires a way to implement strings in Cairo. Strings are also used in ERC-20 (for the optional but so convenient `symbol` and `name` methods) and we have seen projects using them to generate dynamic NFT images (returning the NFT as svg).

In discord we mentioned that strings can be implemented by storing bytes in felts (251 bits or 31 bytes per felt). One issue is that, since StarkNet storage cannot store arrays, that's also the maximum string length that can be stored.

**Comment by martriay** [comment]
Important topic indeed. I'm moving this onto github discussions since I believe this should be implemented at the Cairo level, but keeping the discussion open because I think gathering design input on feature requirements is still valuable.

**Comment by bbrandtom** [comment]
Support for short strings was added in the latest cairo version (v0.5.0) - https://www.cairo-lang.org/docs/how_cairo_works/consts.html#short-string-literals

A wrapper lib that uses arrays to represent longer strings might work as a near-term solution.

@fracek what do you mean by storage not supporting arrays? It is not supported natively as a storage_var, but you can use the lower level storage read/write functions to implement it.

**Reply by franceccon** [reply]
Do you have an example how to do that? The documentation doesn't have any information about storage read/write.

**Reply by bbrandtom** [reply]
[Here](https://github.com/starkware-libs/cairo-lang/blob/3d33c4e829a87bc3d88cf04ed6a489e788918b8b/src/starkware/starknet/common/syscalls.cairo#L130) you can see the definition of the functions `storage_read` and `storage_write`.

A simple snippet - 

```
from starkware.starknet.common.syscalls import storage_read, storage_write

func write_to_storage_cell{syscall_ptr : felt*}(key : felt, value : felt):
    storage_write(address=key, value=value)
    return()
end

func read_from_storage_cell{syscall_ptr : felt*}(key : felt) -> (res : felt):
    return storage_read(address=key)
end
```

**Reply by franceccon** [reply]
Thanks that is helpful. It looks like StarkNet contracts have `2**251 - 1` (or `PRIME - 1`) felts of storage, which can be accessed using `storage_read` and `storage_write` and passing the felt address. The `@storage_var` decorator uses the variable name and index to compute an unique storage address, but as developers we can use the low level functions to access _any_ storage location.

**Reply by martriay** [reply]
This should definitely be documented!

**Reply by bbrandtom** [reply]
I agree it is missing!
We are now writing documentation to the following things, hopefully it won't take too long - 

- block fields and hash
- txn fields and hash
- Sequencer APIs
- contracts:
  - storage layout
  - contract address
  - function selector
  - storage variables
  - Contract ABI
- chain id values

**Comment by rsodre** [comment]
Longer strings support is coming to Cairo soon.
https://github.com/starkware-libs/cairo/issues/3218

Will the contracts be updated to use it on erc721 `tokenURI`?
Is there any official NFT standard on Starket yet?

**Reply by martriay** [reply]
yes they will be updated once we have native support. there's a few SNIP drafts, and I expect more to come.
